// simply-djs/lib/simplydjs.js
"use strict";
// ------------------------------
// -------- E R R O R S ---------
// ------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.ticketSystem = exports.suggestSystem = exports.stealEmoji = exports.nqn = exports.manageSug = exports.manageBtn = exports.giveawaySystem = exports.embedPages = exports.embedCreate = exports.bumpSystem = exports.automeme = exports.manageSuggest = exports.manageTicket = exports.manageGiveaway = exports.manageBtnRole = exports.tictactoe = exports.ticketSetup = exports.suggest = exports.starboard = exports.rps = exports.menuPages = exports.giveaway = exports.ghostPing = exports.buttonPages = exports.embedCreator = exports.chatbot = exports.connect = exports.calculator = exports.bumpReminder = exports.btnRole = exports.betterBtnRole = exports.meme = exports.toRgb = exports.toButtonStyle = exports.emitError = exports.disableButtons = exports.ms = exports.https = exports.version = void 0;
const error_1 = require("./src/error");
// Get the node version
if (+process.version.slice(1, 3) - 0 < 16)
    throw new error_1.SimplyError({
        function: 'simply-djs [CORE]',
        title: `NodeJS Version 16 or newer is required, but you are using ${process.version}.`,
        tip: `Install nodejs 16 or higher in https://nodejs.org/`
    });
try {
    require('discord.js');
}
catch (e) {
    // Emit a warning if there is no discord.js
    process.emitWarning(new error_1.SimplyError({
        function: 'simply-djs [CORE]',
        title: 'Discord.JS is required for this package to run',
        tip: 'This package is optimized to run with discord.js'
    }));
}
const { version: discordJSVersion } = require(require('path').join(require.resolve('discord.js'), '..', '..', 'package.json'));
// Get Discord.JS version using their package.json
if (Number(discordJSVersion.slice(0, 2)) < 14)
    throw new error_1.SimplyError({
        function: 'simply-djs [CORE]',
        title: `Discord.JS version 14.x.x is required, but you are using ${discordJSVersion}. See https://www.npmjs.com/package/discord.js`,
        tip: 'This package is not optimized for Discord.JS v13 or lower. Please use simply-djs v3.0.2 for Discord.JS v13 support'
    });
// ------------------------------
// ------- E X P O R T S --------
// ------------------------------
exports.version = '4.1.1';
var misc_1 = require("./src/misc");
Object.defineProperty(exports, "https", { enumerable: true, get: function () { return misc_1.https; } });
Object.defineProperty(exports, "ms", { enumerable: true, get: function () { return misc_1.ms; } });
Object.defineProperty(exports, "disableButtons", { enumerable: true, get: function () { return misc_1.disableButtons; } });
Object.defineProperty(exports, "emitError", { enumerable: true, get: function () { return misc_1.emitError; } });
Object.defineProperty(exports, "toButtonStyle", { enumerable: true, get: function () { return misc_1.toButtonStyle; } });
Object.defineProperty(exports, "toRgb", { enumerable: true, get: function () { return misc_1.toRgb; } });
var meme_1 = require("./src/meme");
Object.defineProperty(exports, "meme", { enumerable: true, get: function () { return meme_1.meme; } });
var betterBtnRole_1 = require("./src/betterBtnRole");
Object.defineProperty(exports, "betterBtnRole", { enumerable: true, get: function () { return betterBtnRole_1.betterBtnRole; } });
var btnRole_1 = require("./src/btnRole");
Object.defineProperty(exports, "btnRole", { enumerable: true, get: function () { return btnRole_1.btnRole; } });
var bumpReminder_1 = require("./src/bumpReminder");
Object.defineProperty(exports, "bumpReminder", { enumerable: true, get: function () { return bumpReminder_1.bumpReminder; } });
var calculator_1 = require("./src/calculator");
Object.defineProperty(exports, "calculator", { enumerable: true, get: function () { return calculator_1.calculator; } });
var connect_1 = require("./src/connect");
Object.defineProperty(exports, "connect", { enumerable: true, get: function () { return connect_1.connect; } });
var chatbot_1 = require("./src/chatbot");
Object.defineProperty(exports, "chatbot", { enumerable: true, get: function () { return chatbot_1.chatbot; } });
var embedCreator_1 = require("./src/embedCreator");
Object.defineProperty(exports, "embedCreator", { enumerable: true, get: function () { return embedCreator_1.embedCreator; } });
var buttonPages_1 = require("./src/buttonPages");
Object.defineProperty(exports, "buttonPages", { enumerable: true, get: function () { return buttonPages_1.buttonPages; } });
var ghostPing_1 = require("./src/ghostPing");
Object.defineProperty(exports, "ghostPing", { enumerable: true, get: function () { return ghostPing_1.ghostPing; } });
var giveaway_1 = require("./src/giveaway");
Object.defineProperty(exports, "giveaway", { enumerable: true, get: function () { return giveaway_1.giveaway; } });
var menuPages_1 = require("./src/menuPages");
Object.defineProperty(exports, "menuPages", { enumerable: true, get: function () { return menuPages_1.menuPages; } });
var rps_1 = require("./src/rps");
Object.defineProperty(exports, "rps", { enumerable: true, get: function () { return rps_1.rps; } });
var starboard_1 = require("./src/starboard");
Object.defineProperty(exports, "starboard", { enumerable: true, get: function () { return starboard_1.starboard; } });
var suggest_1 = require("./src/suggest");
Object.defineProperty(exports, "suggest", { enumerable: true, get: function () { return suggest_1.suggest; } });
var ticketSetup_1 = require("./src/ticketSetup");
Object.defineProperty(exports, "ticketSetup", { enumerable: true, get: function () { return ticketSetup_1.ticketSetup; } });
var tictactoe_1 = require("./src/tictactoe");
Object.defineProperty(exports, "tictactoe", { enumerable: true, get: function () { return tictactoe_1.tictactoe; } });
// ------------------------------
// ------ H A N D L E R S -------
// ------------------------------
var manageBtnRole_1 = require("./src/handler/manageBtnRole");
Object.defineProperty(exports, "manageBtnRole", { enumerable: true, get: function () { return manageBtnRole_1.manageBtnRole; } });
var manageGiveaway_1 = require("./src/handler/manageGiveaway");
Object.defineProperty(exports, "manageGiveaway", { enumerable: true, get: function () { return manageGiveaway_1.manageGiveaway; } });
var manageTicket_1 = require("./src/handler/manageTicket");
Object.defineProperty(exports, "manageTicket", { enumerable: true, get: function () { return manageTicket_1.manageTicket; } });
var manageSuggest_1 = require("./src/handler/manageSuggest");
Object.defineProperty(exports, "manageSuggest", { enumerable: true, get: function () { return manageSuggest_1.manageSuggest; } });
// ------------------------------
// ---- D E P R E C A T E D -----
// ------------------------------
var DEP_automeme_1 = require("./src/deprecated/DEP-automeme");
Object.defineProperty(exports, "automeme", { enumerable: true, get: function () { return DEP_automeme_1.automeme; } });
var DEP_bumpSystem_1 = require("./src/deprecated/DEP-bumpSystem");
Object.defineProperty(exports, "bumpSystem", { enumerable: true, get: function () { return DEP_bumpSystem_1.bumpSystem; } });
var DEP_embedCreate_1 = require("./src/deprecated/DEP-embedCreate");
Object.defineProperty(exports, "embedCreate", { enumerable: true, get: function () { return DEP_embedCreate_1.embedCreate; } });
var DEP_embedPages_1 = require("./src/deprecated/DEP-embedPages");
Object.defineProperty(exports, "embedPages", { enumerable: true, get: function () { return DEP_embedPages_1.embedPages; } });
var DEP_giveawaySystem_1 = require("./src/deprecated/DEP-giveawaySystem");
Object.defineProperty(exports, "giveawaySystem", { enumerable: true, get: function () { return DEP_giveawaySystem_1.giveawaySystem; } });
var DEP_manageBtn_1 = require("./src/deprecated/DEP-manageBtn");
Object.defineProperty(exports, "manageBtn", { enumerable: true, get: function () { return DEP_manageBtn_1.manageBtn; } });
var DEP_manageSug_1 = require("./src/deprecated/DEP-manageSug");
Object.defineProperty(exports, "manageSug", { enumerable: true, get: function () { return DEP_manageSug_1.manageSug; } });
var DEP_nqn_1 = require("./src/deprecated/DEP-nqn");
Object.defineProperty(exports, "nqn", { enumerable: true, get: function () { return DEP_nqn_1.nqn; } });
var DEP_stealEmoji_1 = require("./src/deprecated/DEP-stealEmoji");
Object.defineProperty(exports, "stealEmoji", { enumerable: true, get: function () { return DEP_stealEmoji_1.stealEmoji; } });
var DEP_suggestSystem_1 = require("./src/deprecated/DEP-suggestSystem");
Object.defineProperty(exports, "suggestSystem", { enumerable: true, get: function () { return DEP_suggestSystem_1.suggestSystem; } });
var DEP_ticketSystem_1 = require("./src/deprecated/DEP-ticketSystem");
Object.defineProperty(exports, "ticketSystem", { enumerable: true, get: function () { return DEP_ticketSystem_1.ticketSystem; } });
